File: 1176_2.py
from sys import stdin

input = stdin.readline


def solve():
    N, K = map(int, input().split())
    arr = [int(input()) for _ in range(N)]
    # dp[i번째 학생][서있는 학생들 bit]
    total_bit = 1 << N
    # 0이 아니라 -1로 초기화하는 이유는
    # 해당 케이스일 때 가능한 경우의 수가 0인 것과
    # 구분을 해주기 위함
    dp = [[-1] * total_bit for _ in range(N)]

    def dfs(idx, visited):
        if dp[idx][visited] > -1:
            return dp[idx][visited]

        elif visited == total_bit - 1:
            return 1

        ret = 0
        for nxt in range(N):
            if (bitmask := 1 << nxt) & visited == 0 and abs(arr[idx] - arr[nxt]) > K:
                ret += dfs(nxt, visited | bitmask)

        dp[idx][visited] = ret
        return ret

    ans = 0
    for i in range(N):
        ans += dfs(i, 1 << i)

    return ans


if __name__ == '__main__':
    print(solve())

File: 1176_2.js
const fs = require('fs');

function solve() {
	const input = fs.readFileSync('/dev/stdin').toString().split('\n');
	let inputIdx = 0;

	const [N, K] = input[inputIdx++].split(' ').map((v) => +v);
	const arr = [];
	arr.length = N;
	for (let i = 0; i < N; arr[i++] = +input[inputIdx++]);

	const dp = [];
	const totalBit = 1 << N;
	dp.length = N;
	for (let i = 0; i < N; i++) {
		dp[i] = [];
		dp[i].length = totalBit;
		/* 0이 아니라 -1로 초기화하는 이유는
		 * 해당 케이스일 때 가능한 경우의 수가 0인 것과
		 * 구분을 해주기 위함 */
		for (let j = 0; j < totalBit; dp[i][j++] = -1);
	}

	const dfs = (idx, visited) => {
		if (dp[idx][visited] > -1) return dp[idx][visited];
		else if (visited === totalBit - 1) return 1;

		let ret = 0;
		for (let nxt = 0; nxt < N; nxt++) {
			const bit = 1 << nxt;
			if ((bit & visited) === 0 && Math.abs(arr[idx] - arr[nxt]) > K) {
				ret += dfs(nxt, visited | bit);
			}
		}
		dp[idx][visited] = ret;
		return ret;
	};

	let ans = 0;
	for (let i = 0; i < N; i++) {
		ans += dfs(i, 1 << i);
	}

	console.log(ans);
}

solve();
